File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int start = 0, end = 0, max_length = 0;
        unordered_set<char> current_chars;

        while (end < s.length())
        {
            if (current_chars.find(s[end]) == current_chars.end())
            {
                current_chars.insert(s[end]);
                max_length = max(max_length, end - start + 1);
                end++;
            }
            else {
                current_chars.erase(s[start]);
                start++;
            }
        }
        return max_length;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        length = 0
        L = 0
        arr = set()
        
        for R in range(len(s)):
            while s[R] in arr:
                arr.remove(s[L])
                L += 1
            arr.add(s[R])
            length = max(R - L + 1, length)
            
        return length
                